### PR TITLE
fixes for deprecation warnings on julia 0.7

### PR DESCRIPTION
**Pull request by mcreel**
This takes care of perhaps half of the warnings

(see also #108, #110)

### CODE REVIEW

File: src/NLopt.jl [hunk 1]
```diff
@@ -27,15 +27,15 @@ end
...
     global const nlopt_callback_wrapper_ptr =
         cfunction(nlopt_callback_wrapper,
-                  Cdouble, (Cuint, Ptr{Cdouble}, Ptr{Cdouble}, Ptr{Void}))
+                  Cdouble, (Cuint, Ptr{Cdouble}, Ptr{Cdouble}, Ptr{Nothing}))
     global const nlopt_vcallback_wrapper_ptr =
-        cfunction(nlopt_vcallback_wrapper, Void,
+        cfunction(nlopt_vcallback_wrapper, Nothing,
                   (Cuint, Ptr{Cdouble}, Cuint, Ptr{Cdouble}, Ptr{Cdouble},
-                   Ptr{Void}))
+                   Ptr{Nothing}))
```
Comment (stevengj) [inline src/NLopt.jl]: You should use `Ptr{Cvoid}` for these kinds of `ccall` arguments.

File: src/NLopt.jl [hunk 2]
```diff
@@ -128,24 +128,24 @@ const res2sym = Dict{Cenum,Symbol}(FAILURE=>:FAILURE, INVALID_ARGS=>:INVALID_ARG
 ############################################################################
 # wrapper around nlopt_opt type
 
-const _Opt = Ptr{Void} # nlopt_opt
+const _Opt = Ptr{Nothing} # nlopt_opt
 
 # pass both f and o to the callback so that we can handle exceptions
-type Callback_Data
+mutable struct Callback_Data
```
Comment (stevengj) [inline src/NLopt.jl]: `struct` should be fine.

File: src/NLopt.jl [hunk 3]
```diff
@@ -128,24 +128,24 @@ const res2sym = Dict{Cenum,Symbol}(FAILURE=>:FAILURE, INVALID_ARGS=>:INVALID_ARG
...
     o::Any # should be Opt, but see Julia issue #269
 end
 
-type Opt
+mutable struct Opt
     opt::_Opt
 
     # need to store callback data for objective and constraints in
     # Opt so that they aren't garbage-collected.  cb[1] is the objective.
     cb::Vector{Callback_Data}
```
Comment (stevengj) [inline src/NLopt.jl]: This probably should be changed to `Ref{Callback_Data}`

File: src/NLopt.jl [hunk 4]
```diff
@@ -169,7 +169,7 @@ end
 
 Base.unsafe_convert(::Type{_Opt}, o::Opt) = o.opt # for passing to ccall
 
-destroy(o::Opt) = ccall((:nlopt_destroy,libnlopt), Void, (_Opt,), o)
+destroy(o::Opt) = ccall((:nlopt_destroy,libnlopt), Nothing, (_Opt,), o)
```
Comment (stevengj) [inline src/NLopt.jl]: `Cvoid` again

File: src/NLopt.jl [hunk 5]
```diff
@@ -243,7 +243,7 @@ end
 ############################################################################
 # converting error results into exceptions
 
-type ForcedStop <: Exception end
+mutable struct ForcedStop <: Exception end
```
Comment (stevengj) [inline src/NLopt.jl]: `struct` is fine

Comment (stevengj) [reply src/NLopt.jl]: (This code pre-dates the existence of immutable types, which is why it used `type`.)

File: src/NLopt.jl [hunk 6]
```diff
@@ -518,15 +518,15 @@ function optimize!(o::Opt, x::Vector{Cdouble})
...
     end
-    opt_f = Array{Cdouble}(1)
+    opt_f = Array{Cdouble}(undef,1)
     ret = ccall((:nlopt_optimize,libnlopt), Cenum, (_Opt, Ptr{Cdouble},
                                                      Ptr{Cdouble}),
                 o, x, opt_f)
     return (opt_f[1], x, chks(ret))
 end
 
 optimize{T<:Real}(o::Opt, x::AbstractVector{T}) =
```
Comment (stevengj) [inline src/NLopt.jl]: use `AbstractVector{<:Real}` for this sort of thing

File: src/NLopt.jl (hunk 1 above)
Comment (mcreel) [reply src/NLopt.jl]: OK, made that change. Once I'm done with these suggestions, I'll make another PR.

File: src/NLopt.jl (hunk 5 above)
Comment (mcreel) [reply src/NLopt.jl]: ok

File: src/NLopt.jl (hunk 6 above)
Comment (mcreel) [reply src/NLopt.jl]: OK

**Comment by mcreel** [comment]
This second PR makes the changes to the first PR that were suggested, except for one.

**Comment by stevengj** [comment]
Superseded by #114.

**Pull request closed without merging**